codegen/src/test/resources/software/amazon/awssdk/codegen/poet/client/test-endpoint-discovery-async.java

```diff
@@ -0,0 +1,257 @@
+
+            HttpResponseHandler<AwsServiceException> errorResponseHandler = createErrorResponseHandler(protocolFactory,
+                    operationMetadata);
+
+            return clientHandler.execute(new ClientExecutionParams<DescribeEndpointsRequest, DescribeEndpointsResponse>()
+                    .withMarshaller(new DescribeEndpointsRequestMarshaller(protocolFactory)).withResponseHandler(responseHandler)
+                    .withErrorResponseHandler(errorResponseHandler).withInput(describeEndpointsRequest));
+        } catch (Throwable t) {
+            return CompletableFutureUtils.failedFuture(t);
+        }
+    }
+
+    /**
+     * Invokes the TestDiscoveryIdentifiersRequired operation asynchronously.
+     *
+     * @param testDiscoveryIdentifiersRequiredRequest
+     * @return A Java Future containing the result of the TestDiscoveryIdentifiersRequired operation returned by the
+     *         service.<br/>
+     *         The CompletableFuture returned by this method can be completed exceptionally with the following
+     *         exceptions.
+     *         <ul>
+     *         <li>SdkException Base class for all exceptions that can be thrown by the SDK (both service and client).
+     *         Can be used for catch all scenarios.</li>
+     *         <li>SdkClientException If any client side error occurs such as an IO related failure, failure to get
+     *         credentials, etc.</li>
+     *         <li>EndpointDiscoveryTestException Base class for all service exceptions. Unknown exceptions will be
+     *         thrown as an instance of this type.</li>
+     *         </ul>
+     * @sample EndpointDiscoveryTestAsyncClient.TestDiscoveryIdentifiersRequired
+     */
+    @Override
+    public CompletableFuture<TestDiscoveryIdentifiersRequiredResponse> testDiscoveryIdentifiersRequired(
+            TestDiscoveryIdentifiersRequiredRequest testDiscoveryIdentifiersRequiredRequest) {
+        try {
+            JsonOperationMetadata operationMetadata = JsonOperationMetadata.builder().hasStreamingSuccessResponse(false)
+                    .isPayloadJson(true).build();
+
+            HttpResponseHandler<TestDiscoveryIdentifiersRequiredResponse> responseHandler = protocolFactory
+                    .createResponseHandler(operationMetadata, TestDiscoveryIdentifiersRequiredResponse::builder);
+
+            HttpResponseHandler<AwsServiceException> errorResponseHandler = createErrorResponseHandler(protocolFactory,
+                    operationMetadata);
+
+            String key = clientConfiguration.option(AwsClientOption.CREDENTIALS_PROVIDER).resolveCredentials().accessKeyId();
+            URI cachedEndpoint = endpointDiscoveryCache.get(key,
+                    testDiscoveryIdentifiersRequiredRequest.endpointDiscoveryRequest(),
+                    clientConfiguration.option(SdkClientOption.ENDPOINT));
+            clientConfiguration.copy(o -> o.option(SdkClientOption.ENDPOINT, cachedEndpoint));
```

Comment:
This doesn't change the original configuration. Also we don't want to update the client endpoint right as the new computed endpoint is only valid for that operation.

For endpoint trait, I am passing the hostPrefix to CientExecutionParams and modifying the endpoint in [BaseClientHandler](https://github.com/aws/aws-sdk-java-v2/blob/831f9e9c3ba1ca7b8e8898875d7157131545a28e/core/sdk-core/src/main/java/software/amazon/awssdk/core/client/handler/BaseClientHandler.java#L54). But this is causing issues with path as its being modified multiple times. I am thinking of passing the hostprefix to marshallers and resolve the endpoint there before creating the HttpFullRequest.

Reply:
Ah, yes, was on a different version when I was doing this. Will update.